File: Python-Allan/SplayTree.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class SplayTree:
    def __init__(self):
        self.root = None

    def right_rotate(self, x):
        y = x.left
        x.left = y.right
        y.right = x
        return y

    def left_rotate(self, x):
        y = x.right
        x.right = y.left
        y.left = x
        return y

    def splay(self, root, key):
        if root is None or root.key == key:
            return root
        if key < root.key:
            if root.left is None:
                return root
            if key < root.left.key:
                root.left.left = self.splay(root.left.left, key)
                root = self.right_rotate(root)
            elif key > root.left.key:
                root.left.right = self.splay(root.left.right, key)
                if root.left.right is not None:
                    root.left = self.left_rotate(root.left)
            return root if root.left is None else self.right_rotate(root)
        else:
            if root.right is None:
                return root
            if key > root.right.key:
                root.right.right = self.splay(root.right.right, key)
                root = self.left_rotate(root)
            elif key < root.right.key:
                root.right.left = self.splay(root.right.left, key)
                if root.right.left is not None:
                    root.right = self.right_rotate(root.right)
            return root if root.right is None else self.left_rotate(root)

    def insert(self, key):
        if self.root is None:
            self.root = Node(key)
            return
        self.root = self.splay(self.root, key)
        if self.root.key == key:
            return
        new_node = Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
        else:
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
        self.root = new_node

    def search(self, key):
        self.root = self.splay(self.root, key)
        return self.root is not None and self.root.key == key

    def delete(self, key):
        if self.root is None:
            return
        self.root = self.splay(self.root, key)
        if self.root.key != key:
            return
        if self.root.left is None:
            self.root = self.root.right
        else:
            temp = self.root.right
            self.root = self.splay(self.root.left, key)
            self.root.right = temp

File: C++-Denis/SplayTree.cc
#include <iostream>
using namespace std;

/*
    Aluno: Denis Almeida Ferreira
    Email: denis.almeida.ferreira@ccc.ufcg.edu.br
    Matrícula: 123210975
    Usuário no GitHub: DenisAlmeidaFerreira

    Implementação de uma Árvore Splay
*/

class SplayTree {
private:
    struct Node {
        int value;
        Node *left, *right, *father;
        Node(int val) : value(val), left(nullptr), right(nullptr), father(nullptr) {}
    };
    
    Node* root;
    
    void rotateLeft(Node* node) {
        Node* right = node->right;
        if (!right) return;
        node->right = right->left;
        if (right->left) right->left->father = node;
        right->father = node->father;
        if (!node->father) root = right;
        else if (node == node->father->left) node->father->left = right;
        else node->father->right = right;
        right->left = node;
        node->father = right;
    }
    
    void rotateRight(Node* node) {
        Node* left = node->left;
        if (!left) return;
        node->left = left->right;
        if (left->right) left->right->father = node;
        left->father = node->father;
        if (!node->father) root = left;
        else if (node == node->father->right) node->father->right = left;
        else node->father->left = left;
        left->right = node;
        node->father = left;
    }
    
    void splay(Node* node) {
        while (node->father) {
            Node* father = node->father;
            Node* grandpa = father->father;
            if (!grandpa) {
                if (node == father->left) rotateRight(father);
                else rotateLeft(father);
            } else if ((node == father->left) == (father == grandpa->left)) {
                if (node == father->left) {
                    rotateRight(grandpa);
                    rotateRight(father);
                } else {
                    rotateLeft(grandpa);
                    rotateLeft(father);
                }
            } else {
                if (node == father->left) {
                    rotateRight(father);
                    rotateLeft(grandpa);
                } else {
                    rotateLeft(father);
                    rotateRight(grandpa);
                }
            }
        }
        root = node;
    }
    
    Node* insertRec(Node* node, int value) {
        if (!node) return new Node(value);
        if (value > node->value) {
            if (!node->right) {
                node->right = new Node(value);
                node->right->father = node;
                return node->right;
            }
            return insertRec(node->right, value);
        } else if (value < node->value) {
            if (!node->left) {
                node->left = new Node(value);
                node->left->father = node;
                return node->left;
            }
            return insertRec(node->left, value);
        }
        return node;
    }
    
    Node* searchRec(Node* node, int value) {
        if (!node || node->value == value) return node;
        if (value < node->value) return searchRec(node->left, value);
        return searchRec(node->right, value);
    }
    
    void transplant(Node* u, Node* v) {
        if (!u->father) root = v;
        else if (u == u->father->left) u->father->left = v;
        else u->father->right = v;
        if (v) v->father = u->father;
    }
    
    Node* findMin(Node* node) {
        while (node->left) node = node->left;
        return node;
    }
    
    void printTreeRec(Node* node, int level) {
        if (!node) return;
        printTreeRec(node->right, level + 1);
        for (int i = 0; i < level; i++) cout << "   ";
        cout << node->value << endl;
        printTreeRec(node->left, level + 1);
    }
    
public:
    SplayTree() : root(nullptr) {}
    
    void insert(int value) {
        Node* newNode = insertRec(root, value);
        if (newNode) splay(newNode);
    }
    
    void remove(int value) {
        Node* nodeToRemove = search(value);
        if (!nodeToRemove) return;
        if (!nodeToRemove->left) transplant(nodeToRemove, nodeToRemove->right);
        else if (!nodeToRemove->right) transplant(nodeToRemove, nodeToRemove->left);
        else {
            Node* successor = findMin(nodeToRemove->right);
            if (successor->father != nodeToRemove) {
                transplant(successor, successor->right);
                successor->right = nodeToRemove->right;
                successor->right->father = successor;
            }
            transplant(nodeToRemove, successor);
            successor->left = nodeToRemove->left;
            successor->left->father = successor;
        }
        if (nodeToRemove->father) splay(nodeToRemove->father);
        delete nodeToRemove;
    }
    
    Node* search(int value) {
        Node* node = searchRec(root, value);
        if (node) splay(node);
        return node;
    }
    
    void printTree() {
        printTreeRec(root, 0);
    }
};
